pkg/ruler/compat.go

```diff
@@ -77,52 +92,60 @@ func engineQueryFunc(engine *logql.Engine, overrides RulesLimits, userID string)
 
 // MultiTenantManagerAdapter will wrap a MultiTenantManager which validates loki rules
 func MultiTenantManagerAdapter(mgr ruler.MultiTenantManager) ruler.MultiTenantManager {
-	return &MultiTenantManager{mgr}
+	return &MultiTenantManager{inner: mgr}
 }
 
 // MultiTenantManager wraps a cortex MultiTenantManager but validates loki rules
 type MultiTenantManager struct {
-	ruler.MultiTenantManager
+	inner ruler.MultiTenantManager
+}
+
+func (m *MultiTenantManager) SyncRuleGroups(ctx context.Context, ruleGroups map[string]rulespb.RuleGroupList) {
+	m.inner.SyncRuleGroups(ctx, ruleGroups)
+}
+
+func (m *MultiTenantManager) GetRules(userID string) []*rules.Group {
+	return m.inner.GetRules(userID)
+}
+
+func (m *MultiTenantManager) Stop() {
+	if registry != nil {
+		registry.stop()
+	}
+
+	m.inner.Stop()
 }
 
 // ValidateRuleGroup validates a rulegroup
 func (m *MultiTenantManager) ValidateRuleGroup(grp rulefmt.RuleGroup) []error {
 	return ValidateGroups(grp)
 }
 
-func MemstoreTenantManager(
-	cfg Config,
-	engine *logql.Engine,
-	overrides RulesLimits,
-) ruler.ManagerFactory {
-	var msMetrics *memstoreMetrics
-	var rwMetrics *remoteWriteMetrics
+// MetricsPrefix defines the prefix to use for all metrics in this package
+const MetricsPrefix = "loki_ruler_wal_"
 
-	return ruler.ManagerFactory(func(
+var registry storageRegistry
+
+func MultiTenantRuleManager(cfg Config, engine *logql.Engine, overrides RulesLimits, logger log.Logger, reg prometheus.Registerer) ruler.ManagerFactory {
+	reg = prometheus.WrapRegistererWithPrefix(MetricsPrefix, reg)
+
+	registry = newWALRegistry(log.With(logger, "storage", "registry"), reg, cfg, overrides)
+
+	return func(
 		ctx context.Context,
 		userID string,
 		notifier *notifier.Manager,
 		logger log.Logger,
 		reg prometheus.Registerer,
 	) ruler.RulesManager {
-		// We'll ignore the passed registerer and use the default registerer to avoid prefix issues and other weirdness.
-		// This closure prevents re-registering.
-		registerer := prometheus.DefaultRegisterer
-
-		if msMetrics == nil {
-			msMetrics = newMemstoreMetrics(registerer)
-		}
-
-		if rwMetrics == nil {
-			rwMetrics = newRemoteWriteMetrics(registerer)
-		}
+		registry.configureTenantStorage(userID)
 
 		logger = log.With(logger, "user", userID)
-		queryFunc := engineQueryFunc(engine, overrides, userID)
-		memStore := NewMemStore(userID, queryFunc, msMetrics, 5*time.Minute, log.With(logger, "subcomponent", "MemStore"))
+		queryFunc := engineQueryFunc(engine, overrides, registry, userID)
+		memStore := NewMemStore(userID, queryFunc, newMemstoreMetrics(reg), 5*time.Minute, log.With(logger, "subcomponent", "MemStore"))
 
 		mgr := rules.NewManager(&rules.ManagerOptions{
-			Appendable:      newAppendable(cfg, overrides, logger, userID, rwMetrics),
+			Appendable:      registry,
 			Queryable:       memStore,
```

Comment:
We're still using the `memStore` here as queryable, which is fine, but we are now able to read from Prometheus to determine this with less hacks and remove the memstore implementation altogether. This can be in a later PR though. I opened https://github.com/grafana/loki/issues/4362